File: game-agent-infra/tools/hash_emitter.go
// hash_emitter.go
// EVEZ B1 Cross-Runtime Verification - Go canonical hash emitter
//
// Note: This uses a simplified normalization (no full NFC) since Go's stdlib
// doesn't include NFC normalization. For production, use golang.org/x/text.

package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"sort"
)

// Result for each file
type FileResult struct {
	Hash  string `json:"hash"`
	Error string `json:"error"`
}

// normalizeJSON recursively sorts object keys (simplified normalization)
func normalizeJSON(obj interface{}) interface{} {
	switch v := obj.(type) {
	case string:
		return v
	case []interface{}:
		result := make([]interface{}, len(v))
		for i, elem := range v {
			result[i] = normalizeJSON(elem)
		}
		return result
	case map[string]interface{}:
		keys := make([]string, 0, len(v))
		for k := range v {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		result := make(map[string]interface{})
		for _, k := range keys {
			result[k] = normalizeJSON(v[k])
		}
		return result
	default:
		return v
	}
}

func canonicalHash(path string) (string, error) {
	data, err := ioutil.ReadFile(path)
	if err != nil {
		return "", err
	}

	var obj interface{}
	if err := json.Unmarshal(data, &obj); err != nil {
		return "", err
	}

	normalized := normalizeJSON(obj)
	canonical, err := json.Marshal(normalized)
	if err != nil {
		return "", err
	}

	hash := sha256.Sum256(canonical)
	return "sha256:" + hex.EncodeToString(hash[:]), nil
}

func main() {
	files := os.Args[1:]
	results := make(map[string]FileResult)

	for _, file := range files {
		h, err := canonicalHash(file)
		if err != nil {
			results[file] = FileResult{Hash: "", Error: err.Error()}
		} else {
			results[file] = FileResult{Hash: h, Error: ""}
		}
	}

	output, _ := json.MarshalIndent(results, "", "  ")
	fmt.Println(string(output))
}

File: game-agent-infra/tools/hash_emitter.js
#!/usr/bin/env node
/**
 * Node.js Canonical Hash Emitter — B1 Cross-Runtime Companion
 *
 * Purpose: Reads JSON files and emits SHA-256 hashes using the same
 * canonicalization rules as the Python implementation:
 * - NFC string normalization (Unicode Canonical Composed)
 * - sort_keys: true (all object keys sorted lexicographically)
 * - compact JSON (no extra whitespace)
 * - UTF-8 encoding
 * - "sha256:" prefix on hash
 *
 * Usage: node hash_emitter.js file1.json file2.json ...
 */

"use strict";

const { createHash } = require("crypto");
const { readFileSync } = require("fs");

// ─── Canonicalization helpers ──────────────────────────────────────────────────

function normalizeStrings(obj) {
    if (typeof obj === "string") {
        return obj.normalize("NFC");
    }
    if (Array.isArray(obj)) {
        return obj.map(normalizeStrings);
    }
    if (obj !== null && typeof obj === "object") {
        const out = {};
        for (const key of Object.keys(obj).sort()) {
            out[key] = normalizeStrings(obj[key]);
        }
        return out;
    }
    return obj;
}

function canonicalJson(obj) {
    const normalized = normalizeStrings(obj);
    const jsonStr = JSON.stringify(normalized);
    return Buffer.from(jsonStr, "utf-8");
}

function sha256(data) {
    return "sha256:" + createHash("sha256").update(data).digest("hex");
}

// ─── Main hash routine ─────────────────────────────────────────────────────────

function hashFile(filePath) {
    try {
        const rawJson = readFileSync(filePath, "utf-8");
        const obj = JSON.parse(rawJson);
        const hash = sha256(canonicalJson(obj));
        return { hash, error: null };
    } catch (err) {
        return { hash: null, error: err.message };
    }
}

// ─── CLI ───────────────────────────────────────────────────────────────────────

function main() {
    const args = process.argv.slice(2);

    if (args.length === 0) {
        console.error("Usage: node hash_emitter.js <file1.json> [file2.json ...]");
        process.exit(1);
    }

    const results = {};
    for (const filePath of args) {
        results[filePath] = hashFile(filePath);
    }

    console.log(JSON.stringify(results, null, 2));
}

if (require.main === module) {
    main();
}

module.exports = {
    normalizeStrings,
    canonicalJson,
    sha256,
    hashFile,
};